triedb/firewood/database.go

```diff
@@ -0,0 +1,617 @@
+	}
+
+	currentRoot, err := fw.Root()
+	if err != nil {
+		log.Crit("firewood: error getting current root", "error", err)
+	}
+
+	return &Database{
+		fwDisk:      fw,
+		ethdb:       diskdb,
+		proposalMap: make(map[common.Hash][]*ProposalContext),
+		proposalTree: &ProposalContext{
+			Root: common.Hash(currentRoot),
+		},
+	}
+}
+
+func validateConfig(diskdb ethdb.Database, trieConfig *Config) (*ffi.Config, string, error) {
+	// Get the path from the database
+	path, err := customrawdb.ReadChainDataPath(diskdb)
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to read database path: %w", err)
+	}
+
+	// Check that the directory exists
+	info, err := os.Stat(path)
+	if err != nil {
+		return nil, "", fmt.Errorf("error checking database path: %w", err)
+	} else if !info.IsDir() {
+		return nil, "", fmt.Errorf("database path is not a directory: %s", path)
+	}
+
+	// Append the filename to the path
+	if trieConfig.FileName == "" {
+		return nil, "", errors.New("no filename provided")
+	}
+	path = filepath.Join(path, trieConfig.FileName)
+
+	// Check if the file exists
+	info, err = os.Stat(path)
+	exists := false
+	if err == nil {
+		if info.IsDir() {
+			return nil, "", fmt.Errorf("database file path is a directory: %s", path)
+		}
+		// File exists
+		log.Info("Database file found", "path", path)
+		exists = true
+	}
+
+	// Create the Firewood config from the provided config.
+	config := &ffi.Config{
+		Create:            !exists,                               // Use any existing file
+		NodeCacheEntries:  uint(trieConfig.CleanCacheSize) / 256, // TODO: estimate 256 bytes per node
+		Revisions:         trieConfig.Revisions,
+		ReadCacheStrategy: trieConfig.ReadCacheStrategy,
+		MetricsPort:       trieConfig.MetricsPort,
+	}
+
+	return config, path, nil
+}
+
+// Scheme returns the scheme of the database.
+// This is only used in some API calls
+// and in StateDB to avoid iterating through deleted storage tries.
+func (db *Database) Scheme() string {
+	return rawdb.HashScheme
+}
+
+// Initialized indicates whether the root provided is the genesis root of the database.
+func (db *Database) Initialized(root common.Hash) bool {
+	// We store the genesis root in the rawdb, so we can check it.
+	genesisRoot, err := customrawdb.ReadFirewoodGenesisRoot(db.ethdb)
+	if err != nil {
+		log.Error("firewood: error reading genesis root", "error", err)
+		return false
+	}
+	return genesisRoot == root
+}
+
+// Update takes a root and a set of keys-values and creates a new proposal.
+// It will not be committed until the Commit method is called.
+// This function should be called even if there are no changes to the state to ensure proper tracking of block hashes.
+func (db *Database) Update(root common.Hash, parentRoot common.Hash, block uint64, nodes *trienode.MergedNodeSet, _ *triestate.Set, opts ...stateconf.TrieDBUpdateOption) error {
+	db.proposalLock.Lock()
+	defer db.proposalLock.Unlock()
```

Comment:
Is this too early? Can't we wait until we called ExtractTrieDBUpdatePayload?

Also when is this released? Can't we release it while firewood is processing the proposal?